stdlib/public/runtime/SymbolInfo.cpp

```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+static Win32ModuleInfo moduleInfoFromAddress(const void *address) {
+  HMODULE hModule;
+  if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS
+                        | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+                         (LPCTSTR)address,
+                         &hModule)) {
+    MODULEINFO mi;
+
+    if (!GetModuleInformation(GetCurrentProcess(), hModule,
+                              &mi, sizeof(mi))) {
+      ::memset(&mi, 0, sizeof(mi));
+    }
+
+    WCHAR wszBuffer[256];
+    LPWSTR pwszFileName = wszBuffer;
+    DWORD dwCapacity = sizeof(wszBuffer) / sizeof(WCHAR);
+    DWORD dwRet = GetModuleFileNameW(hModule, pwszFileName, dwCapacity);
+
+    if (dwRet == dwCapacity) {
+      dwCapacity = 512;
+
+      pwszFileName = (LPWSTR)::malloc(sizeof(WCHAR) * dwCapacity);
+      while (true) {
+        dwRet = GetModuleFileNameW(hModule, pwszFileName, dwCapacity);
+        if (dwRet != dwCapacity)
+          break;
+
+        dwCapacity *= 2;
+
+        pwszFileName = (LPWSTR)::realloc(pwszFileName,
+                                         sizeof(WCHAR) * dwCapacity);
+      }
+    }
+
+    if (dwRet == 0) {
+      if (pwszFileName != wszBuffer)
+        ::free(pwszFileName);
 
+      return Win32ModuleInfo(::strdup("<unknown>"), mi.lpBaseOfDll);
+    }
+
+    const char *result = _swift_win32_copyUTF8FromWide(pwszFileName);
+
+    if (pwszFileName != wszBuffer)
+      ::free((void *)pwszFileName);
+
+    return Win32ModuleInfo(result, mi.lpBaseOfDll);
+  } else {
+    return Win32ModuleInfo(::strdup("<unknown>"), nullptr);
+  }
+}
+#endif
+
+std::optional<SymbolInfo> SymbolInfo::lookup(const void *address) {
 #if defined(__wasm__)
   // Currently, Wasm doesn't have a standard stable ABI for exporting address <->
   // symbol table, it's work in progress. Also, there is no API to access such
   // information from Wasm binary side. It's accessible only from host VM.
   // See https://github.com/WebAssembly/tool-conventions/blob/main/DynamicLinking.md
 #elif defined(_WIN32) && !defined(__CYGWIN__)
+  Win32ModuleInfo moduleInfo = moduleInfoFromAddress(address);
+  SYMBOL_INFO_PACKAGE package;
+  BOOL bRet;
+
+  package.si.SizeOfStruct = sizeof(SYMBOL_INFO);
+  package.si.MaxNameLen = MAX_SYM_NAME;
+
   _swift_win32_withDbgHelpLibrary([&] (HANDLE hProcess) {
     if (!hProcess) {
-        return;
+      bRet = false;
+      return;
     }
 
-    SymbolInfo info;
-    info._package.si.SizeOfStruct = sizeof(SYMBOL_INFO);
-    info._package.si.MaxNameLen = MAX_SYM_NAME;
-    if (SymFromAddr(hProcess, reinterpret_cast<const DWORD64>(address),
-                    nullptr, &info._package.si)) {
-      result = info;
-    }
+    bRet = SymFromAddr(hProcess, reinterpret_cast<const DWORD64>(address),
+                       nullptr, &package.si);
   });
+
+  if (bRet) {
+    return SymbolInfo((const void *)package.si.Address,
+                      ::strdup(package.si.Name),
```

Comment:
I was trying to avoid the heap allocation here, FWIW, since we might be crashing due to allocation failure.

Reply:
FWIW, we do heap allocation all over the shop in the surrounding code, so I'm presently unconcerned about that. If this was actually for a crash handler, as opposed to `fatalError()`, I'd be more worried, obviously.

Reply:
Fair enough.